tests/Controller/CheckoutReselectingPaymentApiTest.php

```diff
@@ -0,0 +1,82 @@
+<?php
+
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sylius\Tests\Controller;
+
+use Symfony\Component\HttpFoundation\Response;
+
+/**
+ * @author Mateusz Zalewski <mateusz.zalewski@lakion.com>
+ */
+class CheckoutPaymentShippingApiTest extends CheckoutApiTestCase
+{
+    /**
+     * @test
+     */
+    public function it_denies_reselecting_order_payment_for_non_authenticated_user()
+    {
+        $this->client->request('PUT', '/api/checkouts/reselect-payment/1');
+
+        $response = $this->client->getResponse();
+        $this->assertResponse($response, 'authentication/access_denied_response', Response::HTTP_UNAUTHORIZED);
+    }
+
+    /**
+     * @test
+     */
+    public function it_does_not_allow_to_reselect_paymeny_of_unexisting_order()
+    {
+        $this->loadFixturesFromFile('authentication/api_administrator.yml');
+
+        $this->client->request('PUT', '/api/checkouts/reselect-payment/1', [], [], static::$authorizedHeaderWithContentType);
+
+        $response = $this->client->getResponse();
+        $this->assertResponseCode($response, Response::HTTP_NOT_FOUND);
+    }
+
+    /**
+     * @test
+     */
+    public function it_does_not_allow_to_reselect_payment_of_order_that_has_no_payment_method_selected()
+    {
+        $this->loadFixturesFromFile('authentication/api_administrator.yml');
+        $checkoutData = $this->loadFixturesFromFile('resources/checkout.yml');
+
+        $orderId = $checkoutData['order1']->getId();
+        $this->addressOrder($orderId);
+
+        $url = sprintf('/api/checkouts/reselect-payment/%d', $orderId);
+        $this->client->request('PUT', $url, [], [], static::$authorizedHeaderWithContentType);
+
+        $response = $this->client->getResponse();
+        $this->assertResponseCode($response, Response::HTTP_BAD_REQUEST);
+    }
+
+    /**
+     * @test
+     */
+    public function it_allows_to_reselect_paymeny_of_order_with_paymeny_method_selected()
+    {
+        $this->loadFixturesFromFile('authentication/api_administrator.yml');
+        $checkoutData = $this->loadFixturesFromFile('resources/checkout.yml');
+
+        $orderId = $checkoutData['order1']->getId();
+        $this->addressOrder($orderId);
+        $this->selectOrderShippingMethod($orderId, $checkoutData['ups']->getId());
+        $this->selectOrderPaymentMethod($orderId, $checkoutData['cash_on_delivery']->getId());
+
+        $url = sprintf('/api/checkouts/reselect-payment/%d', $orderId);
+        $this->client->request('PUT', $url, [], [], static::$authorizedHeaderWithContentType);
+
+        $response = $this->client->getResponse();
+        $this->assertResponse($response, 'checkout/reselect_payment_response', Response::HTTP_OK);
```

Comment:
Again, it's ``applyStateMachineTransitionAction`` and it always return [HTTP_OK response after successful non-html call](https://github.com/Sylius/Sylius/blob/7653df1267c3d4a6037edacf99e89253779a7111/src/Sylius/Bundle/ResourceBundle/Controller/ResourceController.php#L412)